### ISSUE TITLE
Shell commands not properly parsing output

### ISSUE THREAD

**Issue by impala454** [betwo/vscode-catkin-tools#24]
Currently the [build directory helper function](https://github.com/betwo/vscode-catkin-tools/blob/master/src/catkin_workspace.ts#L455) ends up doing a shell execute of `catkin locate -b` and then assumes the first line of output contains the expected response.  Since this is running a user shell, any extra output a user may have in their initial shell will contaminate this assumption and cause the function to fail.   
   
Example: add to your `.bashrc`:

    echo "this is my shell"

And the extension will error out with ```Build directory this is my shell does not exist```

**Comment by betwo** [comment]
Hi @impala454,

thank you for the report.

Could you please help me in reproducing this on my machines?

### How I tried to reproduce

I added 
```bash
echo "This is my bash"
``` 
into my `.bashrc` and made sure that it is always executed

The output of running `bash -c <cmd>` only gives me the output of `cmd`, without any `echo`'s of `.bashrc`:
```bash
$ bash
This is my bash
$ bash -c "echo test"
test
```

Forcing `bash` to run interactively, I can reproduce the error:
```bash
$ bash -lc "echo test"
This is my bash
test
```

However, the extension calls `bash -c '<cmd>'`, so we should be in the non-interactive case here.

### My Setup
```bash
which bash
/bin/bash
```

```bash
bash --version
GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

### Further debugging

Could it be that somehow your shell is always interactive? 
Are you using `bash` or another shell like `zsh`?

**Comment by impala454** [comment]
Interesting.  It looks like the [code that executes the command](https://github.com/betwo/vscode-catkin-tools/blob/3ef5b6900d245df287d2fe7770c0f1e0962aafad/src/catkin_command.ts#L32) may be opening a shell first.  The [documentation](https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback) for that function says it opens a shell and then executes the command:

> Spawns a shell then executes the command within that shell, buffering any generated output. 

So (disclaimer I am not a nodejs expert) I think the way to reproduce is to make a quick nodejs thing that calls this function and see if it's running your command interactively.  Either way, we removed our `echo` temporarily and it definitely fixed the issue.

 I can also add that this is occurring in a docker container using the [Remote-Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) plugin.  Not sure if that changes how a terminal is executed or not.

**Comment by impala454** [comment]
One more quick add, one of our team members mentioned using `child_process.execFile` instead.  The documentation specifies:

> The child_process.execFile() function is similar to child_process.exec() except that it does not spawn a shell by default.

So that may be what's needed here, though you'd still have to have the user's `/opt/ros/{ROS_DISTRO}/setup.bash` in there.

**Comment by impala454** [comment]
I am having a hard time reproducing this as well, without showing you our huge proprietary company docker, here's a quick example I threw together (which still doesn't exhibit the behavior). https://gitlab.com/cclaunch/node_shell_test.  Will let you know if I can get it.

**Comment by oneironautics** [comment]
Other team member here. I created a small extension to test this. Invoking `exec` inside of my test VSCode extension spawns a shell that sources `.bashrc`, which results in `stdout` containing additional output. Invoking `execFile` provides only the expected output of the command passed to it.

The following is the code for the test extension (I've omitted all of the boilerplate for clarity).

```javascript
const vscode = require('vscode');

function activate(context) {
	let disposable = vscode.commands.registerCommand('shelltest.catkinExecTest', function () {
		const cp = require("child_process");
		const oc = vscode.window.createOutputChannel("catkin exec/execFile test");

		cp.exec("bash -c 'catkin locate -b'", { cwd : "/root/catkin_ws" }, (error, stdout, stderr) => {
			oc.appendLine(`exec error: ${ error }`);
			oc.appendLine(`exec stdout: ${ stdout }`);
			oc.appendLine(`exec stderr: ${ stderr }`);
		});

		cp.execFile("catkin", [ "locate", "-b" ], { cwd : "/root/catkin_ws" }, (error, stdout, stderr) => {
			oc.appendLine(`execFile error: ${ error }`);
			oc.appendLine(`execFile stdout: ${ stdout }`);
			oc.appendLine(`execFile stderr: ${ stderr }`);
		});
	});

	context.subscriptions.push(disposable);
}
exports.activate = activate;

module.exports = {
	activate
}
```

The following is the output generated from running the extension:

```
execFile error: null
execFile stdout: /root/catkin_ws/build

execFile stderr: 
exec error: null
exec stdout: message from .bashrc.d
/root/catkin_ws/build

exec stderr: 
```

**Comment by betwo** [comment]
Thank you both for the updates!

That's so strange, I still cannot reproduce this inside vscode.
I copied the same code into an extension and ran it:

```bash
exec error: null
exec stdout: /lhome/sebbuck/workspace/csapex/build

exec stderr: 
execFile error: null
execFile stdout: /lhome/sebbuck/workspace/csapex/build

execFile stderr: 
```

with my `.bashrc`:
```bash
$ cat .bashrc | head -n 5
# ~/.bashrc: executed by bash(1) for non-login shells.
# see /usr/share/doc/bash/examples/startup-files (in the package bash-doc)
# for examples

echo "This is my bash"
```

and
```bash
$ bash -c "echo test"
test
$ bash -lc "echo test"
This is my bash
test
$ bash -ic "echo test"
This is my bash
test
```

---

We can use `execFile` for the `catkin locate` calls, but I'm not sure about all the other commands that are currently relying on a sourced `setup.<shell>` file. 
Looking into that now.

@oneironautics - Could you please trace a `bash -version` and `which bash` from within your temporary extension? I'd like to understand why your shells seem to be running interactively / as login-shells, are we using fundamentally different setups?
Could you also try running `bash --norc -c ...` once in your extension and see if that would suppress the sourcing of `.bashrc` as a quick fix?

**Comment by oneironautics** [comment]
```
bash -version: GNU bash, version 4.4.20(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
--------------------------------
which bash: /bin/bash
--------------------------------
exec error: null
exec stdout: message from .bashrc.d
/root/aquanaut_ws/build

exec stderr: 
--------------------------------
exec (norc) error: null
exec (norc) stdout: /root/aquanaut_ws/build

exec (norc) stderr: 
--------------------------------
execFile error: null
execFile stdout: /root/aquanaut_ws/build

execFile stderr: 
```

It appears that `--norc` has the desired effect.

**Comment by betwo** [comment]
Thank you for confirming this.

There is now a pre-release v1.19.0 that integrates your feedback:
1. Calls to `catkin` now use `execFile`, since no `setup.<shell>` files need to be sourced for them,
2. The shell type is now a configuration parameter (supported are `bash`, `sh` or `zsh`),
3. `bash` and `sh` are called with the option `--norc`.

It would be amazing if you could quickly test this to check if your problems are solved.
You can either pull the latest master, or use the `.vsix` file attached:
[b2-catkin-tools-1.19.0.zip](https://github.com/betwo/vscode-catkin-tools/files/5313930/b2-catkin-tools-1.19.0.zip)

**Comment by oneironautics** [comment]
Looks fixed to me - I no longer see an error message using the provided 1.19.0 `.vsix`.

**Comment by betwo** [comment]
Thank you both for your help! 
I just released v1.19.0, should be available in a few minutes.

**Comment by oneironautics** [comment]
Sure thing. Thanks for the extension!
